core/src/com/unciv/logic/map/BFS.kt

```diff
@@ -1,59 +1,80 @@
 package com.unciv.logic.map
 
-// Kotlin's ArrayDeque is experimental
-import java.util.ArrayDeque
+import kotlin.collections.ArrayDeque
 
 /**
- * Defines intermediate steps of a breadth-first search, for use in either get shortest path or get onnected tiles.
+ * Defines intermediate steps of a breadth-first search, for use in either get shortest path or get connected tiles.
  */
-class BFS(val startingPoint: TileInfo, val predicate : (TileInfo) -> Boolean) {
-    var tilesToCheck = ArrayDeque<TileInfo>()
+class BFS(
+    val startingPoint: TileInfo,
+    private val predicate : (TileInfo) -> Boolean
+) {
+    /** remaining tiles to check */
+    private val tilesToCheck = ArrayDeque<TileInfo>(37)  // needs resize at distance 4
 
     /** each tile reached points to its parent tile, where we got to it from */
-    val tilesReached = HashMap<TileInfo, TileInfo>()
+    private val tilesReached = HashMap<TileInfo, TileInfo>()
 
     init {
         tilesToCheck.add(startingPoint)
         tilesReached[startingPoint] = startingPoint
     }
 
+    /** Process fully until there's nowhere left to check */
     fun stepToEnd() {
         while (!hasEnded())
             nextStep()
     }
 
-
+    /**
+     * Process until either [destination] is reached or there's nowhere left to check
+     * @return `this` instance for chaining
+     */
     fun stepUntilDestination(destination: TileInfo): BFS {
         while (!tilesReached.containsKey(destination) && !hasEnded())
             nextStep()
         return this
     }
 
+    /**
+     * Process one tile-to-search, fetching all neighbors not yet touched
+     * and adding those that fulfill the [predicate] to the reached set
+     * and to the yet-to-be-processed set.
+     * 
+     * Will do nothing when [hasEnded] returns `true`
+     */
     fun nextStep() {
-        val current = tilesToCheck.remove()
```

Comment:
Am I missing something or is the current tile no longer removed from the set of tiles to check?

Reply:
Damn. In my tested java variant that was a `pollFirst`, and that translates to `removeFirstOrNull`... The shame...